src/vs/workbench/contrib/positronNotebook/browser/markdownRenderer.ts

```diff
@@ -86,11 +113,118 @@ function markedHighlight(options: marked.MarkedOptions & {
+	return token.type === 'text';
+}
+
+/**
+ * Type guard to check if a token is a Code token.
+ * @param token The token to check
+ * @returns True if the token is a Code token
+ */
+function isCodeToken(token: marked.Token): token is Extract<marked.Token, { type: 'code' }> {
+	return token.type === 'code';
+}
+
+/**
+ * Token types with a required nested tokens array.
+ * Extracted from marked.Token for type safety.
+ */
+type TokenWithRequiredNestedTokens = Extract<marked.Token, { tokens: marked.Token[] }>;
+
+/**
+ * String literal union of token types with required nested tokens.
+ */
+type RequiredNestedTokenType = TokenWithRequiredNestedTokens['type'];
+
+/**
+ * Set of token types that have nested tokens.
+ * The `satisfies` ensures compile-time validation - if marked adds/removes
+ * token types, the compiler will catch mismatches.
+ */
+const NESTED_TOKEN_TYPES = new Set([
+	'heading',
+	'blockquote',
+	'paragraph',
+	'list_item',
+	'link',
+	'strong',
+	'em',
+	'del'
+] as const satisfies readonly RequiredNestedTokenType[]);
+
+/**
+ * Type guard to check if a token has nested tokens by checking token type.
+ * Uses type discriminator properties instead of the 'in' operator.
+ * @param token The token to check
+ * @returns True if the token is a type that has nested tokens
+ */
+function hasNestedTokens(token: marked.Token): token is Extract<marked.Token, { tokens?: marked.Token[] }> {
+	if (NESTED_TOKEN_TYPES.has(token.type as RequiredNestedTokenType)) {
+		return true;
+	}
+	// For Generic tokens or other tokens, check if tokens property exists using hasOwnProperty
+	// This avoids using the 'in' operator while still checking for the property
+	if (Object.prototype.hasOwnProperty.call(token, 'tokens')) {
+		const genericToken = token as marked.Tokens.Generic;
+		return Array.isArray(genericToken.tokens);
+	}
+	return false;
+}
+
+/**
+ * Extracts plain text from a token array (used for headings).
+ * Recursively processes tokens to extract their text content.
+ * @param tokens Array of tokens to extract text from
+ * @returns Plain text string extracted from tokens
+ */
+function extractTextFromTokens(tokens: marked.Token[]): string {
+	const parts: string[] = [];
+	for (const token of tokens) {
+		if (isTextToken(token)) {
+			parts.push(token.text);
+		} else if (isCodeToken(token)) {
+			parts.push(token.text);
+		} else if (hasNestedTokens(token) && token.tokens !== undefined) {
+			// Recursively process nested tokens
+			parts.push(extractTextFromTokens(token.tokens));
+		}
+	}
+	return parts.join('');
+}
+
+/**
+ * Slugifies text to create URL-safe IDs for headings.
+ * Converts text to lowercase, replaces spaces with hyphens, and removes
+ * punctuation characters. Uses Unicode-aware normalization and property escapes
+ * for better international character support.
+ * @param text Text to slugify
+ * @returns URL-safe slug string
+ */
+function slugify(text: string): string {
```

Comment:
Non-blocking: I was wondering if there was an existing `slugify` we could use and found the public `markedGfmHeadingIdPlugin` function. Could we use that or does it differ/do we prefer the control this gives us?